flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java

```diff
@@ -358,26 +364,57 @@ public static boolean stateBackendFromApplicationOrConfigOrDefaultUseManagedMemo
         return false;
     }
 
+    /**
+     * Load state backend which may wrap the original state backend for recovery.
+     *
+     * @param originalStateBackend StateBackend loaded from application or config.
+     * @param classLoader User code classloader.
+     * @param keyedStateHandles The state handles for restore.
+     * @return Wrapped state backend for recovery.
+     * @throws DynamicCodeLoadingException Thrown if keyed state handles of wrapped state backend
+     *     are found and the class was not found or could not be instantiated.
+     */
+    public static StateBackend loadStateBackendFromKeyedStateHandles(
+            StateBackend originalStateBackend,
+            ClassLoader classLoader,
+            Collection<KeyedStateHandle> keyedStateHandles)
+            throws DynamicCodeLoadingException {
+        // Wrapping ChangelogStateBackend or ChangelogStateBackendHandle is not supported currently.
+        if (!isChangelogStateBackend(originalStateBackend)
+                && keyedStateHandles.stream()
+                        .anyMatch(
+                                stateHandle ->
+                                        stateHandle instanceof ChangelogStateBackendHandle)) {
+            return loadChangelogStateBackend(
+                    originalStateBackend, classLoader, CHANGELOG_STATE_BACKEND_FOR_RECOVERY);
+        }
+        return originalStateBackend;
+    }
+
+    private static boolean isChangelogStateBackend(StateBackend backend) {
+        return CHANGELOG_STATE_BACKEND.equals(backend.getClass().getName());
+    }
+
     private static StateBackend loadChangelogStateBackend(
-            StateBackend backend, ClassLoader classLoader) throws DynamicCodeLoadingException {
+            StateBackend backend, ClassLoader classLoader, String className)
```

Comment:
NIT: now the method is not specific to changelog (but to `DelegatingStateBackend`). So it can be renamed to something like `wrapStateBackend`